### PR TITLE
hc can now handle bundles

### PR DESCRIPTION
**Pull request by freesig**
Follows on from #665 to allow hc to handle bundles and register dnas.
[Demo video](https://youtu.be/rUUGUoKZvDo)

### CODE REVIEW

File: crates/hc_bundle/src/cli.rs [hunk 1]
```diff
@@ -117,7 +121,9 @@ impl HcDnaBundle {
                 crate::init::init_dna(path).await?;
             }
             Self::Pack { path, output } => {
-                let (bundle_path, _) = crate::packing::pack::<DnaManifest>(&path, output).await?;
+                let name = get_dna_name(&path).await?;
+                let (bundle_path, _) =
+                    crate::packing::pack::<DnaManifest>(&path, output, name).await?;
```
Comment (maackle) [inline crates/hc_bundle/src/cli.rs:126]: Kinda weird that we have to read the manifest file twice now, but not a big deal, and I can't think of any immediate suggestion that would be obviously better and not more work than it's work

File: crates/hc_sandbox/src/bundles.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+use walkdir::WalkDir;
+
+#[derive(Debug, Clone)]
+/// Either a set of Dnas or a hApp bundle.
+pub enum DnasHapp {
+    /// Set of Dnas to install.
+    Dnas(Vec<PathBuf>),
+    /// Happ bundle to install.
+    HApp(Option<PathBuf>),
+}
```
Comment (maackle) [inline crates/hc_sandbox/src/bundles.rs:16]: This type isn't used anywhere, is there a future plan for it, or can this file be removed?

File: crates/hc_sandbox/src/bundles.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+impl DnasHapp {
+    /// Create a new DnasHapp from either dnas or happ bundle.
+    pub fn new(dnas: Option<Vec<PathBuf>>, happ: Option<Option<PathBuf>>) -> Self {
+        match (dnas, happ) {
+            (Some(dnas), None) => Self::Dnas(dnas),
+            (None, Some(happ)) => Self::HApp(happ),
+            _ => unreachable!("Cannot have both dnas and happ"),
```
Comment (maackle) [inline crates/hc_sandbox/src/bundles.rs:24]: If this type is being used, and the arguments come straight from the commandline, is this definitely unreachable? I guess only if using structopt "mutually exclusive" args or whatever.

**Review by maackle — APPROVED** [review]
Looks good other than maybe removing `DnasHapp`

File: crates/hc_sandbox/src/bundles.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    let mut happ = happ.unwrap_or(std::env::current_dir()?);
+    if happ.is_dir() {
+        let file_path = search_for_happ(&happ)?;
+        happ = file_path;
+    }
+    ensure!(
+        happ.file_name()
+            .map(|f| f.to_string_lossy().ends_with(".happ"))
+            .unwrap_or(false),
+        "File {} is not a valid dna file name: (e.g. my-dna.dna)",
```
Comment (Connoropolous) [inline crates/hc_sandbox/src/bundles.rs:81]: my-happ.happ would make sense here instead wouldn't it

**Comment by guillemcordoba** [comment]
Not enough expertise to do a review, but I've bumped into an issue: when I rerun the sandbox for the second time after having created it, my UI refuses to connect to it. Here is what I'm doing: 

```bash
hc s create --root sandbox/
hc s call register-dna --path grapes.dna.workdir/grapes-test.dna
hc s call install-app uhC0kIPXFPDB2XSCUcSNLFSSLoBhXFFIRm1aBmMQE9KLD2Lk7LWd1
hc s run --ports 8888
```

Here, I run my UI and it connects fine with the conductor. But then, I kill the `hc s run --ports 8888` and run it again, and now the UI doesn't connect, as if holochain wasn't attached to the port - which seems to be confirmed by `netstat`. Here is the output from the terminal:

```bash
[nix-shell:~/projects/holochain/compository/grapes/zome]$ RUST_LOG=info hc s run --ports 8888
Mar 08 10:56:14.900  INFO holochain_cli_sandbox::run: 

----
starting holochain
----


Mar 08 10:56:14.959  INFO holochain::conductor::conductor::builder: self.config=ConductorConfig { environment_path: EnvironmentRootPath("sandbox/qXIzmgiB-HqQDOMHXG77J"), use_dangerous_test_keystore: false, signing_service_uri: None, encryption_service_uri: None, decryption_service_uri: None, dpki: None, keystore_path: Some("sandbox/qXIzmgiB-HqQDOMHXG77J/keystore"), passphrase_service: None, admin_interfaces: Some([AdminInterfaceConfig { driver: Websocket { port: 0 } }]), network: None }
Mar 08 10:56:15.183  WARN holochain_lmdb::transaction: long-lived reader: 192 ms
Mar 08 10:56:15.212  INFO holochain::conductor::conductor: ADD CELL cell_id=CellId(DnaHash(uhC0kIPXFPDB2XSCUcSNLFSSLoBhXFFIRm1aBmMQE9KLD2Lk7LWd1), AgentPubKey(uhCAkc6TSICVEq4axpWlMg53fCAIRVnqzshabDfNAkypcgrsw-oHk))

hc-sandbox: Running conductor on admin port 43473
hc-sandbox: Attaching app port 8888
Mar 08 10:56:15.214  INFO holochain: Conductor successfully initialized.
Conductor ready.
```

So not seeing any error. 

Cleaning the sandbox and starting the process over again gives the same result: working first time, not working second time.

Happy to convert this into an issue after the PR is merged if this is more appropriate.

File: crates/hc_bundle/src/cli.rs (hunk 1 above)
Comment (freesig) [reply crates/hc_bundle/src/cli.rs:126]: Yeh it's not ideal. Also don't want to do more work than it's work :)

**Comment by freesig** [comment]
@guillemcordoba Thanks for pointing this out.
So adding app interfaces is a persistent operation. If you do `hc s run --ports 8888` and then hc s run` port 8888 will still be attached.
I have fixed it though so there's better error messaging for this now.

**Comment by guillemcordoba** [comment]
So awesome @freesig lots of thanks!

**Pull request merged**